File: src/statesman/commands/PrintCommand.java
package statesman.commands;

import java.util.Random;

import statesman.Content;

public class PrintCommand extends Command {

    public static final String ID_PRINT = "print";
    public static final String ID_PRINTC = "printc";
    public static final String ID_PRINTR = "printr";

    private enum PrintType {
        SINGLE, CONCATENATE, RANDOM
    }

    private boolean _initialized;
    private String[] _messages;

    private PrintType _printType;
    private Random _random;

    private PrintCommand() {
        _initialized = false;
        _messages = null;
        _printType = PrintType.SINGLE;
    }

    public PrintCommand(String[] messages, PrintType printType) {
        this();
        _messages = messages;
        _printType = printType;
        if (printType == PrintType.RANDOM) {
            _random = new Random();
        }
    }

    private static Command _defaultInstance;

    public static Command getDefault() {
        if (_defaultInstance == null) {
            _defaultInstance = new PrintCommand();
        }
        return _defaultInstance;
    }

    @Override
    public void execute() {
        initializeStrings();

        if (_printType == PrintType.CONCATENATE) {
            for (int i = 0; i < _messages.length; i++) {
                System.out.printf(_messages[i]);
            }
            return;
        }

        int i = 0;
        if (_printType == PrintType.RANDOM) {
            i = _random.nextInt(_messages.length);
        }
        System.out.printf(_messages[i] + "%n");
    }

    @Override
    public Command fromText(String commandId, String[] arguments) {
        if (arguments.length < 2) {
            return null;
        }

        String[] messages = new String[arguments.length - 1];
        System.arraycopy(arguments, 1, messages, 0, arguments.length - 1);

        PrintType printType;
        switch (commandId) {
        case ID_PRINT:
            printType = PrintType.SINGLE;
            break;
        case ID_PRINTC:
            printType = PrintType.CONCATENATE;
            break;
        case ID_PRINTR:
            printType = PrintType.RANDOM;
            break;
        default:
            return null;
        }

        return new PrintCommand(messages, printType);
    }

    private void initializeStrings() {
        if (_initialized) {
            return;
        }

        for (int i = 0; i < _messages.length; i++) {
            _messages[i] = Content.getScript().findMessage(
                    _messages[i], false);
        }
        _initialized = true;
    }

}

File: src/Commands/PrintCommand.cs
namespace Statesman.Commands
{
    public class PrintCommand : Command
    {
        public enum PrintType
        {
            Single,
            Concatenate,
            Random
        }

        public const string kIdPrintSingle = "print";
        public const string kIdPrintConcatenate = "printc";
        public const string kIdPrintRandom = "printr";

        protected bool _initialized;
        protected string[] _messages;

        private readonly PrintType _printType;
        private readonly Random _random;

        public PrintCommand(string[] messages, PrintType printType)
        {
            _initialized = false;
            _messages = null;
            _messages = messages;
            _printType = printType;
            if (printType == PrintType.Random)
            {
                _random = new Random();
            }
        }

        public static Command FromText(string commandId, string[] arguments)
        {
            if (arguments.Length < 2)
            {
                return null;
            }

            string[] messages = new string[arguments.Length - 1];
            Array.Copy(arguments, 1, messages, 0, arguments.Length - 1);

            PrintType printType;
            switch (commandId)
            {
                case kIdPrintSingle:
                    printType = PrintType.Single;
                    break;
                case kIdPrintConcatenate:
                    printType = PrintType.Concatenate;
                    break;
                case kIdPrintRandom:
                    printType = PrintType.Random;
                    break;
                default:
                    return null;
            }

            return new PrintCommand(messages, printType);
        }

        public override void Execute()
        {
            InitializeStrings();

            if (_printType == PrintType.Concatenate)
            {
                for (int currentIndex = 0;
                     currentIndex < _messages.Length;
                     currentIndex++)
                {
                    Console.Write(_messages[currentIndex]);
                }
                return;
            }

            int index = 0;
            if (_printType == PrintType.Random)
            {
                index = _random.Next(_messages.Length);
            }
            Console.WriteLine(_messages[index]);
        }

        protected void InitializeStrings()
        {
            if (!_initialized)
            {
                for (int i = 0; i < _messages.Length; i++)
                {
                    _messages[i] = Content.Script.FindMessage(_messages[i], false);
                }
                _initialized = true;
            }
        }
    }
}
